### DISCUSSION TITLE
`CodeEditor`/`CodeBlock` control

### DISCUSSION THREAD

**Discussion by Lamparter** [CommunityToolkit/Labs-Windows#573] [Ideas]
![image](https://github.com/user-attachments/assets/7feef341-6505-4811-b9ca-ffedbdb74a21)

Hey there!
Was recently thinking about replacing the editor control in Yair's [Quick Pad](https://github.com/yaira2/Quick-Pad) project with something like [`WinUIEditor`](https://github.com/BreeceW/WinUIEdit) or a [Monaco](https://github.com/microsoft/monaco-editor)-style editor. The problem with the first is that it's not very actively maintained and I myself don't know C++ so I couldn't continue development on it through a fork; the problem with the second being that it isn't quite so consistent with the WinUI3 design style spec. So it struck me, what better to have a control that does this, similar to `TextBlock` or `TextBox` but with syntax highlighting, Roslyn analysers and other features (see the mock-up).

**Comment by dongle-the-gadget** [comment]
I don't think we can have too advanced syntax highlighting features here (i.e. Roslyn and LSPs), since that will contribute to a lot of code bloat, especially if only a few languages are used. In addition, they are very slow to run and might not be sandbox-friendly.

**Comment by michael-hawker** [comment]
I had written one of the original Monaco wrappers back for UWP, it sort of worked with WinUI. Others have also tried. It's very hard to host the WebView and have it be fully featured. It's a project in itself. (Sorry for the delay in responding here, missed this original post.)

[WinUIEdit](https://github.com/BreeceW/WinUIEdit) is by far the most native performant solution I've come across since then. Being built on a stable base of the existing [Scintilla editor control](https://www.scintilla.org/) is pretty huge. It's really just glue than having to do everything from scratch. I was blown away by how smooth it was.

@Lamparter the developer is on the [Windows App Community](https://discord.gg/eBHZSKG) discord server, so you could try and coordinate there to see what their plans are or coordinate helping or how to help the project.
